From cb7cbf20bab139cccfc06bae069ac125a5d57787 Mon Sep 17 00:00:00 2001
From: Michael Schuerig <michael@schuerig.de>
Date: Sat, 4 Apr 2009 18:46:41 +0200
Subject: [PATCH 1/2] Load an item only if any of its properties required for display are missing.

---
 grid/DataGrid.js |   24 ++++++++++++++++++++++--
 1 files changed, 22 insertions(+), 2 deletions(-)

diff --git a/grid/DataGrid.js b/grid/DataGrid.js
index 39e7ad7..89b867e 100644
--- a/grid/DataGrid.js
+++ b/grid/DataGrid.js
@@ -231,6 +231,26 @@ dojo.declare("dojox.grid.DataGrid", dojox.grid._Grid, {
 	onFetchError: function(err, req){
 	},
 
+	_structureChanged: function(){
+      this.inherited(arguments);
+      var requiredFields = [];
+      dojo.forEach(this.layout.cells, function(c){
+        if (!c.hidden) {
+          requiredFields.push(c.field);
+        }
+      });
+      this._requiredFields = requiredFields;
+    },
+
+    _isItemLoaded: function(item){
+      for (var prop in this._requiredFields){
+        if (!item.hasProperty(prop)){
+          return false;
+        }
+      }
+      return true;
+    },
+
 	_fetch: function(start, isRender){
 		var start = start || 0;
 		if(this.store && !this._pending_requests[start]){
@@ -255,7 +275,7 @@ dojo.declare("dojox.grid.DataGrid", dojox.grid._Grid, {
 					// Load them if we need to
 					var waitCount = 0;
 					dojo.forEach(items, function(i){
-						if(!store.isItemLoaded(i)){ waitCount++; }
+						if(!this._isItemLoaded(i)){ waitCount++; }
 					});
 					if(waitCount === 0){
 						this._onFetchComplete(items, req);
@@ -267,7 +287,7 @@ dojo.declare("dojox.grid.DataGrid", dojox.grid._Grid, {
 							}
 						};
 						dojo.forEach(items, function(i){
-							if(!store.isItemLoaded(i)){
+							if(!this._isItemLoaded(i)){
 								store.loadItem({item: i, onItem: onItem, scope: this});
 							}
 						}, this);
-- 
1.6.2.4

